src/code/z_rumble.c

```diff
@@ -0,0 +1,135 @@
+/**
+ * @file z_rumble.c
+ *
+ * This file implements an interface for the game state to set up, manage and request use of the rumble pak. Despite
+ * some parts of the system appearing to accommodate all four controller ports, only controller 1 can be instructed
+ * to rumble.
+ * This file is half of the system that runs on the graph thread alongside the game state. The rest of the system that
+ * processes the requests runs on the padmgr thread and is implemented in `sys_rumble.c`.
+ *
+ * @see sys_rumble.c
+ *
+ * @note Original filename is likely z_vibrate.c or similar as it is ordered after z_ss_sram.c and before z_view.c
+ */
+#include "global.h"
+
+RumbleMgr sRumbleMgr;
+
+/**
+ * Padmgr callback to update the state of rumble on Vertical Retrace.
+ *
+ * Unlike every other function in this file, this runs on the padmgr thread.
+ */
+void Rumble_Update(PadMgr* padMgr, void* arg) {
+    RumbleMgr_Update(&sRumbleMgr);
+    PadMgr_RumbleSet(padMgr, sRumbleMgr.rumbleEnable);
+}
+
+/**
+ * Forces the rumble state to use the supplied parameters.
+ * The parameters are the same as in `Rumble_Request`.
+ *
+ * @see Rumble_Request
+ */
+void Rumble_Override(f32 distSq, u8 sourceStrength, u8 duration, u8 decreaseRate) {
+    s32 dist;
+    s32 strength;
+
+    if (distSq > SQ(1000)) {
+        dist = 1000;
+    } else {
+        dist = sqrtf(distSq);
+    }
+
+    if (dist < 1000 && sourceStrength != 0 && decreaseRate != 0) {
+        // Decrease the strength linearly with distance
+        strength = sourceStrength - (dist * 255) / 1000;
+
+        if (strength > 0) {
+            // Note: sRumbleMgr is a shared resource between the graph and padmgr threads, no locking is done
+            // to ensure that the entire request is written before it is possibly used.
+            sRumbleMgr.overrideStrength = strength;
+            sRumbleMgr.overrideDuration = duration;
+            sRumbleMgr.overrideDecreaseRate = decreaseRate;
+        }
+    }
+}
+
+/**
+ * Submits a request to the rumble manager with the properties given in the arguments. If there is no free request slot
+ * the request is silently dropped.
+ *
+ * @param distSq
+ *  Squared distance, usually taken to be from an apparent source to the player in world coordinates.
```

Comment:
I suggest using 4 (or at least 2) spaces for this type of indentation in arguments, so it's a bit easier to read